integration-tests/tests/src/tests/objects.ts

```diff
@@ -17,244 +17,1297 @@
+      let calls = 0;
+      let resolve: any;
+      let reject: any;
+
+      let promise = new Promise((res, rej) => {
+        (resolve = res), (reject = rej);
+      });
+      const obj = this.realm.write(() => {
+        return this.realm.create<IStringOnly>(StringOnlySchema.name, { stringCol: "foo" });
+      });
+
+      obj.addListener((obj: IStringOnly, changes) => {
+        try {
+          calls++;
+          switch (calls) {
+            case 1:
+              break;
+            case 2:
+              expect(changes.deleted).to.be.false;
+              expect(changes.changedProperties.length).equals(1);
+              expect(changes.changedProperties[0]).equals("stringCol");
+              expect(obj["stringCol"]).equals("bar");
+              break;
+            case 3:
+              expect(changes.deleted).to.be.true;
+              expect(changes.changedProperties.length).equals(0);
+              this.realm.close();
+          }
+          resolve();
+        } catch (e) {
+          reject(e);
+        }
+      });
+      await promise;
+
+      promise = new Promise((res, rej) => {
+        (resolve = res), (reject = rej);
+      });
+      this.realm.write(() => {
+        obj["stringCol"] = "bar";
       });
+      await promise;
 
-      expect(john._id.equals(_id)).equals(true);
-      expect(john.name).equals("John Doe");
-      expect(john.age).equals(42);
+      promise = new Promise((res, rej) => {
+        (resolve = res), (reject = rej);
+      });
+      this.realm.write(() => {
+        this.realm.delete(obj);
+      });
+      await promise;
+    });
+    it("can add and remove listeners", async function (this: Mocha.Context & RealmContext) {
+      const obj = this.realm.write(() => {
+        return this.realm.create<IStringOnly>(StringOnlySchema.name, { stringCol: "foo" });
+      });
+
+      let calls = 0;
+      let resolve: any;
+
+      const listener = function (obj: IStringOnly) {
+        calls++;
+        if (calls === 1) {
+          expect(obj["stringCol"]).equals("foo");
+        } else if (calls === 2) {
+          expect(obj["stringCol"]).equals("bar");
+        } else {
+          throw new Error("Too many calls to listener");
+        }
+        resolve();
+      };
 
-      realm.write(() => {
-        realm.create(PersonWithId, { _id, age: 43 }, Realm.UpdateMode.All);
+      const initialNotification = new Promise((r) => (resolve = r));
+      obj.addListener(listener);
+      await initialNotification;
+
+      const firstChange = new Promise((r) => (resolve = r));
+      this.realm.write(() => {
+        obj["stringCol"] = "bar";
       });
+      await firstChange;
+      obj.removeListener(listener);
 
-      expect(john._id.equals(_id)).equals(true);
-      expect(john.name).equals("John Doe");
-      expect(john.age).equals(43);
+      this.realm.write(function () {
+        obj["stringCol"] = "foobar";
+      });
+      await sleep(500);
```

Comment:
I would like to avoid using sleeps, but instead await a promise that resolves when the event fires. This would be a good candidate for a refactor in the future.